README.md
# FitBit OAuth 2.0 Provider

[![Source Code](http://img.shields.io/badge/source-jmitchell38488/oauth2--fitbit-blue.svg?style=flat-square)](https://github.com/jmitchell38488/oauth2-fitbit)
[![Latest Version](https://img.shields.io/github/release/jmitchell38488/oauth2-fitbit.svg?style=flat-square)](https://github.com/jmitchell38488/oauth2-fitbit/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](https://github.com/jmitchell38488/oauth2-fitbit/blob/master/LICENSE)
[![Total Downloads](https://img.shields.io/packagist/dt/league/oauth2-fitbit.svg?style=flat-square)](https://packagist.org/packages/league/oauth2-fitbit)

This package makes it simple to integrate your application with the [FitBit OAuth 2.0](https://dev.fitbit.com/docs/oauth2/) service provider.

## Installation

```
composer require jmitchell38488/oauth2-fitbit
```

## Usage
The FitBit provides two different methods for authenticating with the OAuth 2.0 service, an 
[authorization grant flow](https://dev.fitbit.com/docs/oauth2/#authorization-code-grant-flow) and an 
[implicit grant flow](https://dev.fitbit.com/docs/oauth2/#implicit-grant-flow). Both require different
configuration when instantiating the provider and the implicit grant flow will require once less step.

FitBit also uses a different Authorization header than is provided by the parent library. When
a user authenticates with the FitBit 2.0 API, they need to set Authorization: Basic to generate
the access token, and provide the Authorization header with each subsequent request, however
using Bearer instead of Basic.

Included in the package are three concrete provider classes and an abstract provider
class. The abstract provider class provides shared functionality for the Authorization
and Implicit implementations. The __FitBit__ class extends the __Authorization__ 
class, so you can use that instead of the __Authorization__ class if you prefer. It is 
there for clarity when making authenticated requests. In any case, if you are supporting
either Implicit or Authorization grant flows, you will need to keep track of which
one you've used to authenticate a session, since one will timeout and you can
refresh it, whereas the other will require a user to re-authorize once it has 
timed out.

## Authorization Grant Flow
#### Authenticate session
```php
session_start();
use Jmitchell38488\OAuth2\Client\Provider\FitBitAuthorization;
require_once __DIR__ . '/vendor/autoload.php';

$provider = new FitBitAuthorization([
    'clientId'      => $my_client_id_from_fitbit,
    'clientSecret'  => $my_client_secret_from_fitbit,
    'redirectUri'   => $my_callback_url,
]);

// 1st step: Has the user authorised yet?
if (!isset($_SESSION['oauth2state'])) {
    $authorizationUrl = $provider->getAuthorizationUrl([
        'prompt' => FitBitAuthorization::PROMPT_CONSENT,
        'response_type' => FitBitAuthorization::RESPONSETYPE_CODE,
        'scope' => $provider->getAllScope(),
    ]);
    
    // Set the session state to validate in the callback
    $_SESSION['oauth2state'] = $provider->getState();
    
    header('Location: ' . $authorizationUrl);
    exit;
    
// 2nd step: User has authorised, now lets get the refresh & access tokens
} else if (isset($_GET['state']) && $_GET['state'] == $_SESSION['oauth2state'] && isset($_GET['code']) && !isset($_SESSION['fitbit']['oauth'])) {
    try {
        $token = base64_encode(sprintf('%s:%s', $my_client_id_from_fitbit, $my_client_secret_from_fitbit));
        $accessToken = $provider->getAccessToken('authorization_code', [
            'code'  => $_GET['code'],
            'access_token' => $_GET['code'],
            'token' => $token,
        ]);
        
        unset($_SESSION['oauth2state']);
        $_SESSION['fitbit']['oauth2'] = array(
            'accessToken' => $accessToken->getToken(),
            'expires' => $accessToken->getExpires(),
            'refreshToken' => $accessToken->getRefreshToken(),
        );
    } catch (Exception $ex) {
        print $ex->getMessage();
    }

// 3rd step: Authorised, have tokens, but session needs to be refreshed
} else if (time() > $_SESSION['fitbit']['oauth2']['expires']) {
    try {
        $token = base64_encode(sprintf('%s:%s', $my_client_id_from_fitbit, $my_client_secret_from_fitbit));
        $accessToken = $provider->getAccessToken('refresh_token', [
            'grant_type'    => FitBitAuthorization::GRANTTYPE_REFRESH,
            'access_token'  => $_SESSION['fitbit']['oauth2']['accessToken'],
            'refresh_token'  => $_SESSION['fitbit']['oauth2']['refreshToken'],
            'token'         => $token,
        ]);

        unset($_SESSION['oauth2state']);
        $_SESSION['fitbit']['oauth2'] = array(
            'accessToken' => $accessToken->getToken(),
            'expires' => $accessToken->getExpires(),
            'refreshToken' => $accessToken->getRefreshToken(),
        );
    } catch (Exception $ex) {
        print $ex->getMessage();
    }
}
```
## Implicit Grant Flow
#### Authenticate session
```php
session_start();
use Jmitchell38488\OAuth2\Client\Provider\FitBitImplicit;
require_once __DIR__ . '/vendor/autoload.php';

$provider = new FitBitImplicit([
    'clientId'      => $my_client_id_from_fitbit,
    'clientSecret'  => $my_client_secret_from_fitbit,
    'redirectUri'   => $my_callback_url,
]);

// 1st step: Has the user authorised yet? Or do we need to refresh?
if (!isset($_SESSION['oauth2state'])) {
    $authorizationUrl = $provider->getAuthorizationUrl([
        'prompt' => FitBitImplicit::PROMPT_CONSENT,
        'response_type' => FitBitImplicit::RESPONSETYPE_TOKEN,
        'scope' => $provider->getAllScope(),
        'expires_in' => FitBitImplicit::EXPIRES_IN_DAY // This can be set to 1, 7 or 30 days
    ]);
    
    // Set the session state to validate in the callback
    $_SESSION['oauth2state'] = $provider->getState();
    
    header('Location: ' . $authorizationUrl);
    exit;
    
// 2nd step: User has authorised, now lets get the refresh & access tokens
// The return URL uses fragments, so you will need to implement front-end logic to redirect the 
// user back to the server with the relevant information, since the URL will look like:
// my_callback_uri#scope=nutrition+weight+location+social+heartrate+settings+sleep+activity+profile&state=abcdef1234567890&user_id=ABC123&token_type=Bearer&expires_in=86400&access_token=abcdefghijklmnopqrstuvwxyz1234567890abcdefghijklmnopqrstuvwxyz1234567890abcdefghijklmnopqrstuvwxyz1234567890
} else if (isset($_GET['state']) && $_GET['state'] == $_SESSION['oauth2state'] && isset($_GET['access_token']) && !isset($_SESSION['fitbit']['oauth'])) {
    unset($_SESSION['oauth2state']);
    $_SESSION['fitbit']['oauth2'] = array(
        'accessToken' => $_GET['access_token'],
        'expires' => $_GET['expires_in'],
        'refreshToken' => null,
    );
} 
```

## Making requests
The API endpoints can be found in either the [official API docs](https://dev.fitbit.com/docs)
or the [API explorer](https://apigee.com/me3/embed/console/fitbit?apig_cc=1).

It's important to use the FitBit class intead of the grant flow classes, because
FitBit API requires that you use the Bearer token in the Authorization header, 
rather than the Basic token. If you don't use the FitBit class, the API will
return a 401 unauthorized error.

### To make a request
```php

$endpoint = $provider->getBaseApiUrl() . "user/-/profile." . FitBit::FORMAT_JSON;
$provider = new FitBit([
    'clientId'      => $my_client_id_from_fitbit,
    'clientSecret'  => $my_client_secret_from_fitbit,
    'redirectUri'   => $my_callback_url,
]);

$request = $provider->getAuthenticatedRequest(
    FitBit::METHOD_GET,
    $endpoint,
    $_SESSION['fitbit']['oauth2']['accessToken']
);

$response = $provider->getResponse($request);
```

CHANGELOG.md
# FitBit OAuth 2.0 Client Changelog

## 0.1.1

 _Released: 2015-11-12_
 
 * Added phpdoc documentation
 * Added additional constants to the AbstractFitBit class
 * Changed license and package namespace

## 0.1.0

_Released: 2015-11-11_

* Initial release.


src/Provider/FitBit.php
<?php
/**
 * This file is part of the jmitchell38488/oauth2-fitbit library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Justin Mitchell <jmitchell38488@gmail.com>
 * @license http://opensource.org/licenses/MIT MIT
 * @link https://packagist.org/packages/Jmitchell38488/oauth2-fitbit Packagist
 * @link https://github.com/jmitchell38488/oauth2-fitbit GitHub
 */

namespace Jmitchell38488\OAuth2\Client\Provider;

use League\OAuth2\Client\Tool\BearerAuthorizationTrait;

/**
 * <p>Represents a FitBit OAuth 2.0 service provider.</p>
 * <p>This extends from the
 * <b>FitBitAuthorization</b> class to uses the authorization grant flow to 
 * make requests with the API. You can use either this class, or the 
 * <b>FitBitAuthorization</b> class to make authorization grant flow requests.</p>
 * 
 * <p>This class is inteded to be used after authorization is made, because it
 * uses the <em>Bearer</em> token that's returned from the API authorization, rather
 * than the <em>Basic</em> token that's used as part of the authorization.</p>
 * 
 * @link https://dev.fitbit.com/docs/oauth2/ FitBit API Docs
 * @author Justin Mitchell <jmitchell38488@gmail.com>
 * @since 0.1
 * @package Jmitchell38488\OAuth2\Client\Provider
 */
class FitBit extends FitBitAuthorization
{
    use BearerAuthorizationTrait;

    /**
     * @inheritdoc
     */
    public function __construct(array $options = [], array $collaborators = [])
    {   
        parent::__construct($options, $collaborators);
    }

}

src/Provider/FitBitImplicit.php
<?php
/**
 * This file is part of the jmitchell38488/oauth2-fitbit library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Justin Mitchell <jmitchell38488@gmail.com>
 * @license http://opensource.org/licenses/MIT MIT
 * @link https://packagist.org/packages/Jmitchell38488/oauth2-fitbit Packagist
 * @link https://github.com/jmitchell38488/oauth2-fitbit GitHub
 */

namespace Jmitchell38488\OAuth2\Client\Provider;

use Jmitchell38488\OAuth2\Client\Tool\BasicAuthorizationTrait;
use InvalidArgumentException;

/**
 * <p>Represents a FitBit OAuth 2.0 service provider.</p>
 * <p>This class represents the 
 * Implicit grant flow for FitBit API. Whereas Authorization grant flow requires
 * 4 steps and returns the authorize response as query parameters, the Implicit
 * grant flow requires 3 steps, skipping the token step, and returns the
 * authorize request as a URI fragment rather than query string. A front-end
 * application will need to send the fragment to the server if intending to use
 * for multiple requests.</p>
 *
 * <p>Once a user session is authorized and the access token is returned from
 * the API, you should use the <strong>FitBit</strong> class to make API requests
 * as it uses the <em>Bearer</em> Authorization token, rather than the
 * <em>Basic</em> Authorization token.</p>.
 * 
 * @link https://dev.fitbit.com/docs/oauth2/#implicit-grant-flow Obtaining Consent
 * @author Justin Mitchell <jmitchell38488@gmail.com>
 * @since 0.1
 * @package Jmitchell38488\OAuth2\Client\Provider
 */
class FitBitImplicit extends AbstractFitBit
{
    use BasicAuthorizationTrait;

    /**
     * <p>Returns authorization parameters based on provided options. FitBit 
     * implicit grant flow differs from the generic authorization grant
     * flow in several ways. First, the prompt field is <em>prompt</em> instead of
     * <em>approval_prompt</em>, and the default library will always insert the request
     * field <em>approval_prompt</em>. Second, a scope is required and there is a
     * minimum set of values and a total set of values. If there are any scope
     * errors the API will return an error. Lastly, an <em>expires_in</em> field
     * is required to set the token expiry. The user can override this on the
     * API if the request prompt is set to <em>consent</em></p>
     * 
     * <p><b>Example:</b><br />
     * https://www.fitbit.com/oauth2/authorize?response_type=token&client_id=22942C&redirect_uri=http%3A%2F%2Fexample.com%2Fcallback&scope=activity%20nutrition%20heartrate%20location%20nutrition%20profile%20settings%20sleep%20social%20weight&expires_in=604800
     * </p>
     *
     * @param  array $options
     * @return array Authorization parameters
     * @link https://dev.fitbit.com/docs/oauth2/#authorization-page Authorization
     * @link https://dev.fitbit.com/docs/oauth2/#scope FitBit API Scope
     */
    protected function getAuthorizationParameters(array $options)
    {
        // Check for required scopes
        if (!empty($options['scope']) && count($options['scope']) < count($this->defaultScopes)) {
            throw new InvalidArgumentException("Could not configure provider, missing required scopes");
        }
        
        // If scopes provided but not all required are provided
        if (count(array_intersect($options['scope'], $this->defaultScopes)) != count($this->defaultScopes)) {
            throw new InvalidArgumentException("Could not configure provider, missing required scopes");
        }
        
        // Invalid scope provided
        if (count($options['scope']) - count(array_intersect($options['scope'], $this->allScopes)) != 0) {
            throw new InvalidArgumentException("Could not configure provider, invalid scope(s) provided");
        }
        
        if (empty($options['scope'])) {
            $options['scope'] = $this->defaultScopes;
        }
        
        $poptions = parent::getAuthorizationParameters($options);
        
        // option[approval_prompt] needs to be remapped to [prompt]
        unset($poptions['approval_prompt']);
        if (isset($options['prompt'])) {
            $poptions['prompt'] = $options['prompt'];
        }
        
        // Set the expires query part, if it is not provided, set to 1 day by default
        $poptions['expires_in'] = isset($options['expires_in']) ? $options['expires_in'] : AbstractFitBit::EXPIRES_IN_DAY;
        
        return $poptions;
    }

}

src/Provider/FitBitAuthorization.php
<?php
/**
 * This file is part of the jmitchell38488/oauth2-fitbit library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Justin Mitchell <jmitchell38488@gmail.com>
 * @license http://opensource.org/licenses/MIT MIT
 * @link https://packagist.org/packages/Jmitchell38488/oauth2-fitbit Packagist
 * @link https://github.com/jmitchell38488/oauth2-fitbit GitHub
 */

namespace Jmitchell38488\OAuth2\Client\Provider;

use Jmitchell38488\OAuth2\Client\Tool\BasicAuthorizationTrait;
use InvalidArgumentException;

/**
 * <p>Represents a FitBit Authorization Grant Flow service provider that may be 
 * used to interact with the OAuth 2.0 service provider, using Basic token
 * authentication.</p>
 * 
 * <p>Once a user session is authorized and the access token is returned from
 * the API, you should use the <strong>FitBit</strong> class to make API requests
 * as it uses the <em>Bearer</em> Authorization token, rather than the
 * <em>Basic</em> Authorization token.</p>.
 * 
 * @link https://dev.fitbit.com/docs/oauth2/#authorization-code-grant-flow Obtaining Consent
 * @author Justin Mitchell <jmitchell38488@gmail.com>
 * @since 0.1
 * @package Jmitchell38488\OAuth2\Client\Provider
 */
class FitBitAuthorization extends AbstractFitBit
{
    use BasicAuthorizationTrait;

    /**
     * <p>Returns authorization parameters based on provided options. FitBit 
     * authorization grant flow differs from the generic authorization grant
     * flow in several ways. First, the prompt field is <em>prompt</em> instead of
     * <em>approval_prompt</em>, and the default library will always insert the request
     * field <em>approval_prompt</em>. Second, a scope is required and there is a
     * minimum set of values and a total set of values. If there are any scope
     * errors the API will return an error.</p>
     * 
     * <p><b>Example:</b><br />
     * https://www.fitbit.com/oauth2/authorize?response_type=code&client_id=22942C&redirect_uri=http%3A%2F%2Fexample.com%2Fcallback&scope=activity%20nutrition%20heartrate%20location%20nutrition%20profile%20settings%20sleep%20social%20weight
     * </p>
     *
     * @param  array $options
     * @return array Authorization parameters
     * @link https://dev.fitbit.com/docs/oauth2/#authorization-page Authorization
     * @link https://dev.fitbit.com/docs/oauth2/#scope FitBit API Scope
     */
    protected function getAuthorizationParameters(array $options)
    {
        // Check for required scopes
        if (!empty($options['scope']) && count($options['scope']) < count($this->defaultScopes)) {
            throw new InvalidArgumentException("Could not configure provider, missing required scopes");
        }
        
        // If scopes provided but not all required are provided
        if (count(array_intersect($options['scope'], $this->defaultScopes)) != count($this->defaultScopes)) {
            throw new InvalidArgumentException("Could not configure provider, missing required scopes");
        }
        
        // Invalid scope provided
        if (count($options['scope']) - count(array_intersect($options['scope'], $this->allScopes)) != 0) {
            throw new InvalidArgumentException("Could not configure provider, invalid scope(s) provided");
        }
        
        if (empty($options['scope'])) {
            $options['scope'] = $this->defaultScopes;
        }
        
        $poptions = parent::getAuthorizationParameters($options);
        
        // option[approval_prompt] needs to be remapped to [prompt]
        unset($poptions['approval_prompt']);
        if (isset($options['prompt'])) {
            $poptions['prompt'] = $options['prompt'];
        }
        
        return $poptions;
    }

}

src/Tool/BasicAuthorizationTrait.php
<?php
/**
 * This file is part of the jmitchell38488/oauth2-fitbit library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Justin Mitchell <jmitchell38488@gmail.com>
 * @license http://opensource.org/licenses/MIT MIT
 * @link https://packagist.org/packages/Jmitchell38488/oauth2-fitbit Packagist
 * @link https://github.com/jmitchell38488/oauth2-fitbit GitHub
 */

namespace Jmitchell38488\OAuth2\Client\Tool;

/**
 * Enables `Basic` header authorization for providers. This is required for the 
 * FitBit API
 * 
 * @link https://dev.fitbit.com/docs/oauth2/#obtaining-consent Obtaining Consent
 * @author Justin Mitchell <jmitchell38488@gmail.com>
 * @since 0.1
 * @package Jmitchell38488\OAuth2\Client\Tool
 */
trait BasicAuthorizationTrait
{
    /**
     * <p>Returns authorization headers for the 'basic' grant. The token is 
     * created by base 64 encoding a concatenated string, by combining the 
     * FitBit API Client ID and Client Secret together, separated by a colon.</p>
     * <p><b>Example:</b>
     * 
     * <pre><code>
     * Token: base64_encode(sprintf('%s:%s', $client_id, $client_secret))
     * Return: Y2xpZW50X2lkOmNsaWVudCBzZWNyZXQ=
     * 
     * POST https://api.fitbit.com/oauth2/token
     * Authorization: Basic Y2xpZW50X2lkOmNsaWVudCBzZWNyZXQ=
     * Content-Type: application/x-www-form-urlencoded
     * 
     * client_id=22942C&grant_type=authorization_code&redirect_uri=http%3A%2F%2Fexample.com%2Fcallback&code=1234567890
     * </code></pre>
     *
     * @param  mixed|null $token Either a string or an access token instance
     * @return array
     * @link https://dev.fitbit.com/docs/oauth2/#access-token-request Making a Token Request
     */
    protected function getAuthorizationHeaders($token = null)
    {   
        return ['Authorization' => 'Basic ' . $token];
    }
}


src/Provider/AbstractFitBit.php
<?php
/**
 * This file is part of the jmitchell38488/oauth2-fitbit library
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * @copyright Copyright (c) Justin Mitchell <jmitchell38488@gmail.com>
 * @license http://opensource.org/licenses/MIT MIT
 * @link https://packagist.org/packages/Jmitchell38488/oauth2-fitbit Packagist
 * @link https://github.com/jmitchell38488/oauth2-fitbit GitHub
 */

namespace Jmitchell38488\OAuth2\Client\Provider;

use League\OAuth2\Client\Provider\AbstractProvider;
use League\OAuth2\Client\Token\AccessToken;
use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
use Psr\Http\Message\ResponseInterface;

/**
 * Represents a FitBit OAuth 2.0 service provider. FitBit provides two different
 * grant flows for creating sessions, an authorization grant flow that is
 * intended to be used by servers, and an implicit grant flow that is intended
 * to be used by browsers. Both implementations extend this class.
 * 
 * @link https://dev.fitbit.com/docs/oauth2/ FitBit API Docs
 * @author Justin Mitchell <jmitchell38488@gmail.com>
 * @since 0.1
 * @package Jmitchell38488\OAuth2\Client\Provider
 */
abstract class AbstractFitBit extends AbstractProvider
{
    /**
     * No user prompt on authorization
     */
    const PROMPT_NONE = 'none';
    
    /**
     * The user must login to authorize the request
     */
    const PROMPT_LOGIN = 'login';
    
    /**
     * If the user is already logged in, they only have to consent, otherwise
     * they login and consent
     */
    const PROMPT_CONSENT = 'consent';
    
    /**
     * Required to use the Authorization grant flow
     */
    const RESPONSETYPE_CODE = 'code';
    
    /**
     * Required to use the Implicit grant flow
     */
    const RESPONSETYPE_TOKEN = 'token';
    
    /**
     * Uses the RefreshToken Grant Provider
     */
    const GRANTTYPE_REFRESH = 'refresh_token';
    
    /**
     * Uses the AuthorizationCode Grant Provider
     */
    const GRANTTYPE_AUTH = 'authorization_code';
    
    /**
     * Return in JSON format
     */
    const FORMAT_JSON = 'json';
    
    /**
     * Return in XML format
     */
    const FORMAT_XML = 'xml';
    
    /**
     * For semantics sake in implementing applications that can use the implicit
     * grant flow
     */
    const GRANTFLOW_IMPLICIT = 'implicit';
    
    /**
     * For semantics sake in implementing applications that can use the authorization
     * grant flow
     */
    const GRANTFLOW_AUTHORIZATION = 'authorization';
    
    /**
     * For use in implicit grant flow, authorization expires in 24 hours
     */
    const EXPIRES_IN_DAY = 86400;
    
    /**
     * For use in implicit grant flow, authorization expires in 7 da7s
     */
    const EXPIRES_IN_WEEK = 604800;
    
    /**
     * For use in implicit grant flow, authorization expires in 30 days
     */
    const EXPIRES_IN_MONTH = 2592000;
    
    
    /**
     * A list of required scopes to be provided to FitBit API. This is based on
     * official FitBit developer documents, accessed 13 October 2015.
     * @var array
     */
    protected $defaultScopes = array(
        'activity', 'nutrition', 'profile', 'settings', 'sleep', 'social', 'weight'
    );
    
    /**
     * A list of all scopes available in the FitBit API. This is based on
     * official FitBit developer documents, accessed 13 October 2015.
     * @var array
     */
    protected $allScopes = array(
        'activity', 'heartrate', 'location', 'nutrition', 'profile', 'settings', 
        'sleep', 'social', 'weight'
    );
    
    /**
     * The base URI for FitBit to base all requests from
     * @var string
     */
    protected $urlApiBase;
    
    /**
     * The base URI for the FitBit API Authorization request endpoint
     * @var string
     */
    protected $urlAuthorize;

    /**
     * The base URI for the FitBit API Access Token request endpoint
     * @var string
     */
    protected $urlAccessToken;

    /**
     * The base URI for the FitBit API Resource owner request endpoint, eg. /user/-/profile
     * @var string
     */
    protected $urlResourceOwnerDetails;

    /**
     * @inheritdoc
     */
    protected $accessTokenMethod;

    /**
     * @inheritdoc
     */
    protected $accessTokenResourceOwnerId;

    /**
     * An array of strings representing the selected scopes that you will request
     * authorization to access
     * @var array[string]
     */
    protected $scopes = null;

    /**
     * The scope separator string
     * @var string
     */
    protected $scopeSeparator;

    /**
     * @inheritdoc
     */
    protected $responseError = 'error';

    /**
     * @inheritdoc
     */
    protected $responseCode;

    /**
     * @inheritdoc
     */
    protected $responseResourceOwnerId = 'id';

    /**
     * Constructor for the Abstract FitBit class. The constructor will set the 
     * request URIs that the application will interact with, so that the
     * implementor doesn't need to specify URLs on request.
     * @param array $options An array of options to set on this provider.
     *     Options include `clientId`, `clientSecret`, `redirectUri`, and `state`.
     *     Individual providers may introduce more options, as needed.
     * @param array $collaborators An array of collaborators that may be used to
     *     override this provider's default behavior. Collaborators include
     *     `grantFactory`, `requestFactory`, `httpClient`, and `randomFactory`.
     *     Individual providers may introduce more collaborators, as needed.
     */
    public function __construct(array $options = [], array $collaborators = [])
    {   
        $options = array_merge($options, array(
            'urlApiBase'                => 'https://api.fitbit.com/1/',
            'urlAuthorize'              => 'https://www.fitbit.com/oauth2/authorize',
            'urlAccessToken'            => 'https://api.fitbit.com/oauth2/token',
            'urlResourceOwnerDetails'   => 'https://api.fitbit.com/1/users/-/profile.json',
        ));
        
        parent::__construct($options, $collaborators);
    }

    /**
     * @inheritdoc
     */
    protected function checkResponse(ResponseInterface $response, $data)
    {
        if (!empty($data[$this->responseError])) {
            $error = $data[$this->responseError];
            $code  = $this->responseCode ? $data[$this->responseCode] : 0;
            throw new IdentityProviderException($error, $code, $data);
        }
    }

    /**
     * @inheritdoc
     */
    protected function createResourceOwner(array $response, AccessToken $token)
    {
        return new GenericResourceOwner($response, $this->responseResourceOwnerId);
    }

    /**
     * @inheritdoc
     */
    public function getBaseApiUrl()
    {
        return $this->urlApiBase;
    }

    /**
     * @inheritdoc
     */
    public function getBaseAccessTokenUrl(array $params)
    {
        return $this->urlAccessToken;
    }

    /**
     * @inheritdoc
     */
    public function getBaseAuthorizationUrl()
    {
        return $this->urlAuthorize;
    }

    /**
     * @inheritdoc
     */
    public function getResourceOwnerDetailsUrl(AccessToken $token)
    {
        return $this->urlResourceOwnerDetails;
    }

    /**
     * @inheritdoc
     */
    protected function getScopeSeparator()
    {
        return ' ';
    }

    /**
     * @inheritdoc
     */
    protected function getDefaultScopes() 
    {
        return $this->defaultScopes;
    }
    
    public function getRequiredScope() 
    {
        return $this->defaultScopes;
    }
    
    public function getAllScope() 
    {
        return $this->allScopes;
    }

    /**
     * Returns a prepared request for requesting an access token. In the case of
     * FitBit, we want to send an authenticated request to the API because we
     * need to send an <b>Authorization</b> header. The default method used an
     * unauthenticated request to get the access token.
     *
     * @param array $params Query string parameters
     */
    protected function getAccessTokenRequest(array $params, $token = null)
    {
        $method  = $this->getAccessTokenMethod();
        $url     = $this->getAccessTokenUrl($params);
        $options = $this->getAccessTokenOptions($params);

        return $this->getAuthenticatedRequest($method, $url, $token, $options);
    }

    /**
     * Requests an access token using a specified grant and option set. For FitBit,
     * We're going to pass the response code from the initial authorization
     * request to get the access token. By default this isn't supported because
     * the library makes an unauthenticated request. To make the authenticated
     * request, we need to fetch the code from the options array and pass it
     * to the <em>getAccessTokenRequest</em> method.
     *
     * @param  mixed $grant
     * @param  array $options
     * @return AccessToken
     */
    public function getAccessToken($grant, array $options = [])
    {
        $grant = $this->verifyGrant($grant);

        $params = [
            'client_id'     => $this->clientId,
            'redirect_uri'  => $this->redirectUri,
        ];
        
        $token = null;
        if (!empty($options['token'])) {
            $token = $options['token'];
            unset($options['token']);
        }

        $params   = $grant->prepareRequestParameters($params, $options);
        $request  = $this->getAccessTokenRequest($params, $token);
        $response = $this->getResponse($request);
        $prepared = $this->prepareAccessTokenResponse($response);
        $token    = $this->createAccessToken($prepared, $grant);

        return $token;
    }
    
}